### PR TITLE
Logs: Unify detected fields and labels in Log Details

### LINKED ISSUE #30221: Logs: Unify display of detected-fields/labels
Current UI:
image

**What would you like to be added**:

Allow the same actions to be performed (filter-in, filter-out, show-this-field, ad-hoc statistic) for all fields regardless of whether they are labels or parsed fields. 

**Why is this needed**:

When viewing logs you don't care about the distinction. Currently there's no way to show the value of a label instead of the log message without altering the query, which is very limiting.

Filter-in/out could be more difficult for parsed fields, but an approximate solution using `|=` and `!=` would be fine as a starting point.

### PR DESCRIPTION
**Pull request by svennergr**
**What is this feature?**

With this feature we are unifying detected fields and labels in the details of a log row. This will initially result in less items in Loki, as "js generated" fields are dropped. Those will be available in the log details, if the according parser is added.

**Which issue(s) does this PR fix?**:

Fixes #30221

# Release notice breaking change

Removes support for "detected fields" in the details of a log line, however all supported interactions (filter, statistics, visibility) are now supported for all fields. If you are using Loki you can get those fields back by using a parser operation like `logfmt` or `json`.

### CODE REVIEW

File: public/app/features/logs/components/LogDetails.tsx [hunk 1]
```diff
@@ -111,53 +115,71 @@ class UnThemedLogDetails extends PureComponent<Props> {
...
-                        <Icon
-                          name="question-circle"
-                          size="xs"
-                          className={css`
-                            margin-left: ${theme.spacing(0.5)};
-                          `}
-                        />
-                      </Tooltip>
+                    <td colSpan={100} className={style.logDetailsHeading} aria-label="Data Links">
+                      Data Links
```
Comment (matyax) [inline public/app/features/logs/components/LogDetails.tsx:154]: As far as I understand, we would be removing the concept of _Detected fields_ and adding _Data Links_. Would there be any chance for users to either miss _Detected fields_ and/or be confused by the concept of _Data Links_?

What I mean is that if this change would be undestandable/expected or we need to do any UX/further clarification of its meaning.

File: public/app/features/logs/components/LogDetailsRow.test.tsx [hunk 2]
```diff
@@ -109,21 +92,22 @@ describe('LogDetailsRow', () => {
...
-    const adHocStatsButton = screen.getByTitle('Ad-hoc statistics');
+    const adHocStatsButton = screen.getByRole('button', { name: 'Ad-hoc statistics' });
     fireEvent.click(adHocStatsButton);
     expect(screen.getByTestId('logLabelStats')).toBeInTheDocument();
     expect(screen.getByTestId('logLabelStats')).toHaveTextContent('another value');
   });
 
-  it('should render clipboard button on hover of log row table value', async () => {
-    setup({ parsedKey: 'key', parsedValue: 'value' });
+  // TODO: skipped for now because I don't know a way of testing CSS styles
```
Comment (matyax) [inline public/app/features/logs/components/LogDetailsRow.test.tsx:101]: If it requires testing of the presence or absense of a given CSS class or style, I would recommend to remove the test, and consider using e2e tests for this assertion if it's important. You could use a `data-` attribute for the test, but it doesn't look like it's worth the extra code.

**Comment by matyax** [comment]
I still need to do a thorough review of the changes, but left some comments in the mean time.

File: public/app/features/logs/components/LogDetails.tsx (hunk 1 above)
Comment (svennergr) [reply public/app/features/logs/components/LogDetails.tsx:154]: I don't think users will be confused by _Data Links_, as those are only links (derived fields) which previously would end up in _Detected fields_. 

After some chats with @niat22 and @stevendungan we agreed on monitoring the confusion/frustration. However we could add a small note for some duration, that _Detected Fields_ can be brought back by simply adding the right parser to your query. I am not decided on that, as we already suggest parsers in the query builder.

Comment (svennergr) [reply public/app/features/logs/components/LogDetails.tsx:154]: However we definitely shouldn't forget to change documentation https://grafana.com/docs/grafana/latest/explore/logs-integration/.

Added a FU issue: https://github.com/grafana/grafana/issues/61026

File: public/app/features/logs/components/logParser.ts [hunk 3]
```diff
@@ -94,30 +47,28 @@ const getDataframeFields = memoizeOne(
...
-  return fieldA.key > fieldB.key ? 1 : fieldA.key < fieldB.key ? -1 : 0;
-}
-
 function shouldRemoveField(field: Field, index: number, row: LogRowModel) {
   // Remove field if it is:
   // "labels" field that is in Loki used to store all labels
   if (field.name === 'labels' && field.type === FieldType.other) {
     return true;
   }
   // "id" field which we use for react key
```
Comment (matyax) [inline public/app/features/logs/components/logParser.ts:56]: Is this comment still relevant?

**Review by matyax — APPROVED** [review]
Looks good! Left some comments.

File: public/app/features/logs/components/LogDetailsRow.tsx [hunk 4]
```diff
@@ -155,99 +202,132 @@ class UnThemedLogDetailsRow extends PureComponent<Props, State> {
...
+        ></ToolbarButton>
       ) : (
-        <IconButton name="eye" title="Show this field instead of the message" onClick={this.showField} />
+        <ToolbarButton
+          tooltip="Show this field instead of the message"
+          iconOnly
+          narrow
+          icon="eye"
+          onClick={this.showField}
+        ></ToolbarButton>
```
[the remark this answers is not in the document]
Comment (svennergr) [reply public/app/features/logs/components/LogDetailsRow.tsx]: Good suggestion. Changed in 006848e

File: public/app/features/logs/components/LogRowMessageDetectedFields.tsx [hunk 5]
```diff
@@ -34,6 +34,10 @@ class UnThemedLogRowMessageDetectedFields extends PureComponent<Props> {
           return `${parsedKey}=${field.value}`;
         }
 
+        if (row.labels[parsedKey]) {
```
[the remark this answers is not in the document]
Comment (svennergr) [reply public/app/features/logs/components/LogRowMessageDetectedFields.tsx]: Good catch! Changed that, and renamed the component in 3a2fba2

**Comment by gelicia** [comment]
The solution to unifying two lists is to remove one of the lists? I see there are comments above about monitoring confusion. If this is what has to happen, keeping an eye on a response is good. But we're losing a feature, not gaining one, and I don't think this is what the original ticket had in mind. 

Also, for the "Data Links" field, has there been discussion about unifying this language with "derived fields" in the datasource, so users know where these values come from? 

fields

**Comment by svennergr** [comment]
> The solution to unifying two lists is to remove one of the lists? I see there are comments above about monitoring confusion. If this is what has to happen, keeping an eye on a response is good. But we're losing a feature, not gaining one, and I don't think this is what the original ticket had in mind.

Thanks for the feedback, @gelicia. Most of the answers you can find in the design doc: https://docs.google.com/document/d/1XvZ0KDBG_f7d1fdhGQI6ykMqsK631yvlbegSGbLe__4/edit

But to also answer it here: having both lists was confusing for a lot of people and also not consistent across other logging datasources. With those changes we'll unify the display and interactions to be the same for all of them. If you think about Loki, you will still see the same results if you add the right parser (`json` or `logfmt`) and to help the user with that, we also have solutions in place (like hints).

> Also, for the "Data Links" field, has there been discussion about unifying this language with "derived fields" in the datasource, so users know where these values come from?

This is also very specific for Loki, as those fields are named "derived fields" for Loki, but "data links" for Elasticsearch and other datasources. However I agree that unifying the datasources settings should happen and I know, that there's some work ongoing for that.

File: public/app/features/logs/components/LogRows.test.tsx [hunk 6]
```diff
@@ -20,6 +20,11 @@ describe('LogRows', () => {
         prettifyLogMessage={true}
         timeZone={'utc'}
         enableLogDetails={true}
+        displayedFields={[]}
+        onClickFilterLabel={() => {}}
```
Comment (matyax) [inline public/app/features/logs/components/LogRows.test.tsx:24]: We don't have any assertions for these callbacks?

**Review by matyax — APPROVED** [review]
👌  great work!

**Comment by Bobris** [comment]
I don't know where to place feedback, but removal of Detected Fields is very unfortunate, as we lost very handy feature.
Workaround to always add `| logfmt | __error__=""` is actually even changing query results.

**Comment by honganan** [comment]
> I don't know where to place feedback, but removal of Detected Fields is very unfortunate, as we lost very handy feature. Workaround to always add `| logfmt | __error__=""` is actually even changing query results.

Want `detected fields` back too!

**Comment by matyax** [comment]
Hi @honganan ! We hear your feedback. We're currently working together with the Loki team to address many different scenarios related with fields and labels, including new features. We'd appreciate your patience, and if you have a workflow that has been disrupted, feel free to open an issue to help you find a workaround for it.

**Comment by honganan** [comment]
> Hi @honganan ! We hear your feedback. We're currently working together with the Loki team to address many different scenarios related with fields and labels, including new features. We'd appreciate your patience, and if you have a workflow that has been disrupted, feel free to open an issue to help you find a workaround for it.

That worth to look forward to! Thank you for so fast response! Respect!

**Pull request merged**